### PR TITLE
python3Packages.numpy: use pytestCheckHook & disable failing tests

### PR DESCRIPTION
**Pull request by misuzu**
## Description of changes
Closes: https://github.com/NixOS/nixpkgs/issues/251514

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [X] i686-linux
  - [X] armv7l-linux (native)
  - [X] aarch64-linux
  - [X] x86_64-darwin
  - [X] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/numpy/default.nix [hunk 1]
```diff
@@ -106,10 +106,31 @@ in buildPythonPackage rec {
     typing-extensions
   ];
 
-  checkPhase = ''
+  checkPhase = let
+    # https://github.com/numpy/numpy/issues/24548
+    disabledTests = lib.optionals stdenv.isx86_32 [
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/numpy/default.nix:111]: Whats the difference to isi686 which I would have used. see https://github.com/NixOS/nixpkgs/blob/master/lib/systems/inspect.nix#L20-L21

File: pkgs/development/python-modules/numpy/default.nix [hunk 2]
```diff
@@ -106,10 +106,31 @@ in buildPythonPackage rec {
...
+      "test_big_arrays" # ValueError: array is too big; `arr.size * arr.dtype.itemsize` is larger tha...
+      "test_multinomial_pvals_float32" # Failed: DID NOT RAISE <class 'ValueError'>
+    ] ++ lib.optionals stdenv.isAarch64 [
+      "test_big_arrays" # OOM on a 16G machine
+    ];
+  in ''
     runHook preCheck
     pushd "$out"
-    ${python.interpreter} -c 'import numpy, sys; sys.exit(numpy.test("fast", verbose=10) is False)'
+    pytest -v ${lib.optionalString (disabledTests != [ ]) "-k ${lib.escapeShellArg (lib.concatMapStringsSep " and " (t: "not ${t}") disabledTests)}"}
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/numpy/default.nix:133]: Please use pytestCheckHook

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/numpy/default.nix (hunk 1 above)
Comment (misuzu) [reply pkgs/development/python-modules/numpy/default.nix]: Updated

File: pkgs/development/python-modules/numpy/default.nix (hunk 2 above)
Comment (misuzu) [reply pkgs/development/python-modules/numpy/default.nix]: Done

File: pkgs/development/python-modules/numpy/default.nix [hunk 3]
```diff
@@ -101,19 +101,40 @@ in buildPythonPackage rec {
...
   ];
 
   checkPhase = ''
     runHook preCheck
     pushd "$out"
-    ${python.interpreter} -c 'import numpy, sys; sys.exit(numpy.test("fast", verbose=10) is False)'
+    pytestCheckPhase
     popd
     runHook postCheck
   '';
```
Comment (tjni) [inline pkgs/development/python-modules/numpy/default.nix:115]: Perhaps checking out what they do in their [custom test runner](https://github.com/numpy/numpy/blob/a277f6210739c11028f281b8495faf7da298dbef/numpy/_pytesttester.py#L79) in terms of adding arguments to pytest could help.

1. If we use their existing test runner, it looks like we can pass custom pytest args using the `extra_argv` param.
2. If we switch to pytestCheckHook, do we want to run their full test suite or stick to running their fast test suite?

Comment (misuzu) [reply pkgs/development/python-modules/numpy/default.nix:115]: Disabling test using the first option would be uglier, `pytestCheckHook` is a better option.

> do we want to run their full test suite or stick to running their fast test suite?

Fast one would likely lead to less breakages which imo is proffered. I'll try to replicate it

Comment (misuzu) [reply pkgs/development/python-modules/numpy/default.nix:115]: Done

File: pkgs/development/python-modules/numpy/default.nix [hunk 4]
```diff
@@ -101,19 +101,45 @@ in buildPythonPackage rec {
...
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/numpy/default.nix:115]:
```suggestion
  preCheck = ''
    cd "$out"
  '';
```

**Review by SuperSandro2000 — APPROVED** [review]
lovely, thanks for bearing with us

**Comment by doronbehar** [comment]
Hey all, I was proposing a similar change in https://github.com/NixOS/nixpkgs/pull/255607 , for Numpy and Scipy. I have there now 1 small additional follow-up change to numpy, and another, similar commit for Scipy. I'd be interested to get your review at #255607 and your comments if you have any at https://github.com/NixOS/nixpkgs/issues/255262 .

**Pull request merged**